### ISSUE TITLE
Odd NMS Crash?

### ISSUE THREAD

**Issue by Joannou1** [PaperMC/Paper#681]
This crash has been the bane of my existence this last week.
I can't tell you how many countless hours and sleepless nights have been spent trying to fix it.
I'm writing this out of pure and utter desperation at this point.

Players are given their own worlds where they can pick from a selection of rooms and make a dungeon of sorts.
The worlds are downloaded on join, ran through RegionFixer to be sure they are clean, and loaded.
At some point the server will hang for no apparent reason and eventually crash.
Each instance only has a max of 5 players, so it's not like a lot of worlds are being loaded.
It is a somewhat rare occurance, it must be someones world that has something really nasty that even RegionFixer isn't seeing.
We've tried loading the individual worlds leftover after a crash and seeing if they crashed by themselves, but had no such luck.

The stack is very cryptic, showing some World / BlockState NMS methods.
I have no idea if this is something you guys can fix, I know I tried to make a change in PaperSpigot once and saw the NMS was there and editable (albeit painfully)

Paper version: git-Paper-1094
https://hastebin.com/hiraperojo.md

I'll be happy to provide any paper/spigot configs or anything needed.
I'm out of ideas and out of energy and this is my last ditch, last effort resort.

**Comment by zachbr** [comment]
When the server stops responding for too long, watchdog will kill it. That's all is happening here, it isn't a crash in the normal sense.

When this happens, watchdog will present a stacktrace however oftentimes this stacktrace is _after_ the event that caused the true lag, rather than during. The information presented is not always deterministic of the issue at hand.

I'd recommend you keep track of what's going on in timings preceding these events, as its data will show us more of a historical view. Feel free to share them with us here, best examples are the ones in which the server is under peak load.

**Comment by Joannou1** [comment]
I'm actually having trouble getting timings to work lately, sometimes it will work, but other times timings paste will do nothing despite being turned on 10 or so minutes prior.

However here's possibly some interesting data I might as well drop here.

Performance following the hang (In this particular case, the main thread is 100% hung for last 10 minutes but not triggering the watchdog despite being set at 60 seconds)
I had to run a thread dump from VisualVM to get this.

https://hastebin.com/ekuwagoxez.vbs
![image](https://cloud.githubusercontent.com/assets/997747/25857940/5932f57c-348f-11e7-8ce4-638e244e3035.png)

Here is one timings that we got a couple days ago, although do ignore the libs disguises usage, we've fixed that since (was unrelated to the hanging issue, just misconfigured mythic mob disguise)

https://timings.aikar.co/?id=f74de00c67bc40adba014c7705056914

Timings are pretty good otherwise, a few low count occurrences of world loading and saving.
It's worth noting that sometimes it will try to generate structures despite having that turned off in the spigot config for worlds? (Also disabled in the world generator settings when the world is created)

**Comment by Joannou1** [comment]
Also here's a CPU profile started during the hang
![image](https://cloud.githubusercontent.com/assets/997747/25858481/b36f0ce6-3490-11e7-8d63-1df9395bfc46.png)

http://gld.bz/uploads/snapshot-1494343338279.nps

It seems like the paper watchdog thread is also hung on World.a()?

![image](https://cloud.githubusercontent.com/assets/997747/25858780/6fb0d7f4-3491-11e7-815b-ffc2337d8c0a.png)

**Comment by reverofevil** [comment]
@Joannou1 Isn't it obvious that slow event that kills the server is the process of copying the room?

**Comment by Joannou1** [comment]
@polkovnikov-ph not sure what you mean? It's a very sudden and extreme hang if you look at the graphs.
There's no indication that any other plugin is doing it?
It's also very random, this occurred 5 hours after a much higher than average load, but it can occur with just a few players as well at any time.

**Comment by MrPowerGamerBR** [comment]
@Joannou1 do you have FastAsyncWorldEdit? FAWE sometimes breaks Timings if you don't change "parallel-threads" to 1 in FAWE's config.

**Comment by Joannou1** [comment]
@MrPowerGamerBR I do, but we need the extra processing speed for our game mode, it's mission-critical that the schematics get pasted in as fast as possible.

**Comment by AlfieC** [comment]
Seems like an issue w/ pooled blockpositions & thread safety. Remove FAWE and try again.

**Comment by boy0001** [comment]
Paper is calling it async:
```
java.lang.Exception
	at net.minecraft.server.v1_11_R1.BlockPosition$PooledBlockPosition.e(BlockPosition.java:267)
	at net.minecraft.server.v1_11_R1.BlockPosition$PooledBlockPosition.s(BlockPosition.java:256)
	at net.minecraft.server.v1_11_R1.World.a(World.java:1281)
	at net.minecraft.server.v1_11_R1.World.getCubes(World.java:1332)
	at net.minecraft.server.v1_11_R1.Entity.move(Entity.java:645)
	at net.minecraft.server.v1_11_R1.EntityLiving.g(EntityLiving.java:1787)
	at net.minecraft.server.v1_11_R1.EntityHuman.g(EntityHuman.java:1448)
	at net.minecraft.server.v1_11_R1.EntityLiving.n(EntityLiving.java:2081)
	at net.minecraft.server.v1_11_R1.EntityHuman.n(EntityHuman.java:350)
	at net.minecraft.server.v1_11_R1.EntityLiving.A_(EntityLiving.java:1916)
	at net.minecraft.server.v1_11_R1.EntityHuman.A_(EntityHuman.java:143)
	at net.minecraft.server.v1_11_R1.EntityPlayer.playerTick(EntityPlayer.java:289)
	at net.minecraft.server.v1_11_R1.PlayerList.disconnect(PlayerList.java:408)
	at net.minecraft.server.v1_11_R1.PlayerConnection.a(PlayerConnection.java:1073)
	at net.minecraft.server.v1_11_R1.PlayerConnection.disconnect(PlayerConnection.java:248)
	at org.spigotmc.RestartCommand.restart(RestartCommand.java:58)
	at org.spigotmc.RestartCommand.restart(RestartCommand.java:40)
	at org.spigotmc.WatchdogThread.run(WatchdogThread.java:97)
```

**Comment by zachbr** [comment]
This looks to be related to upstream's [SPIGOT-924](https://hub.spigotmc.org/jira/browse/SPIGOT-924) fix.
Though looking over how the watchdog handles restarts in general, I can't say it was written with thread safety mind. There appear to be plenty of separate issues.

**Comment by zachbr** [comment]
Watchdog thread shouldn't do stupid shit asynchronously now.

**Comment by Shevchik** [comment]
Sync shutdown won't allow server to actually shutdown if it is freezed in a loop i think.

**Comment by zachbr** [comment]
The only way to call sync shutdown is from the main thread, so if it's frozen you won't be calling it.

A call made asynchronously will simply end the tick loop, as the main thread is already frozen and there is no safe way to handle saving or disconnecting players from that state.

**Comment by Shevchik** [comment]
Kill the main thread, do restart in another.

**Comment by zachbr** [comment]
I did not experience any similar issues in testing, and am not sure I fully understand your suggestion.
We already basically do what you're proposing, but rather than a hard sigkill we force the isRunning state of the server to false, ending the tick loop.

Skipping any saving and player ticks that would've been called asynchronously before this.
I do not see any significant advantages to switching to a method that does force a more traditional process kill at this time.

Edit: hmm I think I see your point in a few cases I didn't come across earlier. Appears I overlooked them.

**Comment by zachbr** [comment]
Though I mis-tagged it as 684 rather than 681. I have switched to your suggestion of forcefully killing the main thread in all situations, rather than counting on it being able to do so itself.

I completely overlooked a true deadlock situation in my testing of this change. Thank you for pointing it out @Shevchik.

I think I'm going to stop committing and typing when I'm this tired.

**Comment by Joannou1** [comment]
It seems like the thread is being killed properly now and the server dies peacefully, nice work.
Should I open a new issue for the original issue? (still crashes from time to time with that same odd blockstate thing)

**Comment by electronicboy** [comment]
Depends how it does peacefully, if it's stopping the server and saving the world in the process, that would indicate that your server is basically entering a state where the amount of work for a tick takes just longer than what watchdog allows before killing the server;

Not necessarily a bug with paper, as much as the implications of fast movements which generally cause this, I'm currently not at home so can't trace the stack of your original issue, however; it specifically looks like somebody is moving too fast on the server, which results in large calls to getCubes, which can for various issues cause a timeout of the watchdog... e.g. slow loading of chunks, chunk generation... when somebody is moving fast, it can place lots of strain on these tasks onto the main thread causing a crash in this particularly hot area

**Comment by Joannou1** [comment]
Surely there has to be a way to prevent this from happening regardless?
This issue just struck another gamemode of ours that is completely different than the first one, same getCubes sort of thing except this time the thread dump didn't include the velocity warning.

https://hastebin.com/icimiburad.md

**Comment by electronicboy** [comment]
Well, We could go back to Spigots method of force crashing the server instead in the case of getCubes receiving a large request...

The problem is that there is no real sane way to handle such cases, that section of code is just stupidly expensive, especially in specific cases where players move fast or triggers a large call to that method, you either let it run like it's supposed to, or you introduce issues with trying to stop it from executing for long...

**Comment by Joannou1** [comment]
Hm yeah, I just find it odd that it could run for that long, in the case of my initial post, it was never ending, I let it run for 12+ hours. It just seems like such a request wouldn't be possible unless the velocity was such a truly ridiculously large number that it sent the player flying literally millions or billions of blocks away.
(It is worth noting that the player is usually within an enclosed room 99% of the time when they do get knocked around by something and we haven't had any reports of people getting sent to the moon yet)

Sometimes it doesn't even say that velocity was set at an excessive amount.
Maybe an optional clamp parameter could be added to attempt to prevent this from happening?
Or a debug to see when the players velocity is being set, or perhaps a way to detect when an unrealistic amount of getCubes are called? (that last one might be testing the sanity bit you were talking about)

But yeah, something somewhere is misbehaving and it's a really tricky issue.

**Comment by sgdc3** [comment]
Could you post your plugin list? @Joannou1 I'm having the same issue, maybe we can find the plugin that indirectly causes the crash.

**Comment by Joannou1** [comment]
Sure @sgdc3 .

< redacted after below comment because probably not needed and has stuff that does stuff in it >

...we've worked with the developer of those two to make sure velocities are clamped.

**Comment by sgdc3** [comment]
Ok we found it. MythicMobs is the cause, this morning i disabled it and since then 0 crashes.

**Comment by sgdc3** [comment]
@Xikage (MythicMob's dev)

**Comment by sgdc3** [comment]
~~latest mm dev build fixed the issue~~

**Comment by sgdc3** [comment]
Nope, it crashed again @xikage

**Comment by Joannou1** [comment]
@sgdc3 Could you put your crash report or thread dump into a paste/hastebin and put it here?
We clamped more velocities and the issue is still occurring unfortunately.

https://hastebin.com/anikebimoj.md

**Comment by sgdc3** [comment]
the thread dump is always the same

**Comment by Joannou1** [comment]
@sgdc3 is it the same as mine? It might help if we could see it.

**Comment by sgdc3** [comment]
https://pastebin.com/TTA9gRa9

**Comment by Joannou1** [comment]
Huh, yeah, looks exactly the same, down to the same line numbers.
How often would you say this occurs, and with how many players?

**Comment by sgdc3** [comment]
It happens 3/4 times per day, my average player count is around 60/70

**Comment by Joannou1** [comment]
Yeah, seems to happen around 10-30 times a day for me across 10-20 servers with 5-10 players each.

**Comment by zachbr** [comment]
Now that you've narrowed down the issue to a specific plugin and are working with that author to resolve it, there is no need for this ticket to remain open on Paper's tracker.

If the developer of said plugin has any thoughts, they should feel free to contact us.

**Comment by Joannou1** [comment]
@Zbob750 - I built a paperspigot version with velocity debug, it's too early to tell if this is the issue, but I'm guessing NaN might cause some problems.

![image](https://cloud.githubusercontent.com/assets/997747/26522967/bf3fff74-42c0-11e7-9aab-32d3f3d62d89.png)

**Comment by Joannou1** [comment]
Able to consistently reproduce if I set my velocity to NaN while crouching.

**Comment by zachbr** [comment]
Not a paper bug then.

Report to whichever plugin is responsible. Watchdog is supposed to alert you to this as well, but that's separate from the main issue here.

**Comment by Joannou1** [comment]
@Zbob750 Pretty new when it comes to making changes to any of this, would I grab a copy of Spigot or so and make the fix and submit a PR there?

I'm able to fix it by preventing the velocity from being set to NaN.

**Comment by zachbr** [comment]
If you want to submit it there, yes.

**Comment by sgdc3** [comment]
@Joannou1 yes please, that would be awesome! ;)

**Comment by sgdc3** [comment]
@Joannou1 Could you post your debug code please?

**Comment by Joannou1** [comment]
Sure, basically just put this under setVelocity in CraftEntity:

if (Double.isNaN(vel.getX()) || Double.isNaN(vel.getY()) || Double.isNaN(vel.getZ())){
    System.out.println("Got NaN velocity: " + vel);
}

A basic fix would be:
        if (Double.isNaN(vel.getX())) vel.setX(getVelocity().getX());
        if (Double.isNaN(vel.getY())) vel.setY(getVelocity().getY());
        if (Double.isNaN(vel.getZ())) vel.setZ(getVelocity().getZ());

not sure if it's a good idea to set the velocity to its current velocity, might just want to set it to 0 instead.

**Comment by sgdc3** [comment]
I just compiled a test build, let's see if the crash is gone.

**Comment by Joannou1** [comment]
@sgdc3 how's the crashing? I've had none in the last 72 hours since the patch.

**Comment by sgdc3** [comment]
idem

**Comment by Joannou1** [comment]
"the same" as in.. same as me, no crashes?
Or same as in, still crashing as before?

I submitted this to Spigot, they didn't want to fix, so I submitted to Mojang, which they said would fix.
https://hub.spigotmc.org/jira/browse/SPIGOT-3283
https://bugs.mojang.com/projects/MC/issues/MC-118164

**Comment by sgdc3** [comment]
Spigot fixed it! https://hub.spigotmc.org/stash/projects/SPIGOT/repos/craftbukkit/commits/921ba53556066ee16cb212e564ee9403fb6a73ff

**Comment by sgdc3** [comment]
Still happening on 1.12... I managed to fix it in 1.11, i'll send a PR asap.
